src/scripts/hunt-filter/engine.ts

```diff
@@ -0,0 +1,129 @@
+import { ApiResponse, User } from "../types/hg";
+import { IntakeMessage } from "../types/mhct";
+import { ILogger } from "../util/logger";
+import { IRule, IFilteredRule, IMessageExemption } from "./interfaces";
+import * as ResponseRules from "./responseRules";
+import * as UserRules from "./userRules";
+import * as MessageExemptions from "./messageExemptions";
+import * as MessageRules from "./messageRules";
+
+/**
+ * Uses pluggable rule to validate data before a hunt can be
+ * submitted to the database
+ */
+export class IntakeRejectionEngine {
+    private logger: ILogger;
+    private responseRules: IRule<ApiResponse>[] = [];
+    private userRules: IRule<User>[] = [];
+    private messageRules: IFilteredRule<IntakeMessage>[] = [];
+    private messageExemptions: IMessageExemption[] = [];
+
+    constructor(logger: ILogger) {
+        this.logger = logger;
+        this.initResponseRules();
+        this.initUserRules();
+        this.initMessageRules();
+    };
+
+    public validateResponse(pre: ApiResponse, post: ApiResponse): boolean {
+        return this.responseRules.every(r => {
+            const isValid = r.isValid(pre, post);
+            if (!isValid) {
+                this.logger.debug(`Reponse object invalid due to rule: ${r.name}`);
+            }
+            return isValid;
+        });
+    }
+
+    public validateUser(pre: User, post: User): boolean {
+        return this.userRules.every(r => {
+            let isValid = r.isValid(pre, post);
+            if (!isValid) {
+                this.logger.debug(`User object invalid due to rule: ${r.name}`);
+            }
+            return isValid;
+        });
+    }
+
+    public validateMessage(pre: IntakeMessage, post: IntakeMessage): boolean {
+        // TODO: Refactor? Location is being set to null in stage funcs so it'll fail later
+        if (pre.location === null || post.location === null) {
+            return false;
+        }
+
+        // Run rules. Build set of currently invalid properties
+        // { "location", "stage" }
+        const invalidProperties = new Set<(keyof IntakeMessage)>();
+        for (const rule of this.messageRules) {
+            const valid: boolean = rule.isValid(pre, post);
```

Comment:
Why the explicit type? It shouldn't be necessary IMO

Reply:
It will default to `Set<unknown>` if you don't specify the type. This enforces that only keys of IntakeMessage can be added to the set.